freqtrade/templates/FreqaiExampleStrategy.py

```diff
@@ -0,0 +1,194 @@
+
+        informative[coin + "20sma"] = ta.SMA(informative, timeperiod=20)
+        informative[coin + "21ema"] = ta.EMA(informative, timeperiod=21)
+        informative[coin + "bmsb"] = np.where(
+            informative[coin + "20sma"].lt(informative[coin + "21ema"]), 1, 0
+        )
+        informative[coin + "close_over_20sma"] = informative["close"] / informative[coin + "20sma"]
+
+        informative[coin + "mfi"] = ta.MFI(informative, timeperiod=25)
+
+        informative[coin + "ema21"] = ta.EMA(informative, timeperiod=21)
+        informative[coin + "sma20"] = ta.SMA(informative, timeperiod=20)
+        stoch = ta.STOCHRSI(informative, 15, 20, 2, 2)
+        informative[coin + "srsi-fk"] = stoch["fastk"]
+        informative[coin + "srsi-fd"] = stoch["fastd"]
+
+        bollinger = qtpylib.bollinger_bands(qtpylib.typical_price(informative), window=14, stds=2.2)
+        informative[coin + "bb_lowerband"] = bollinger["lower"]
+        informative[coin + "bb_middleband"] = bollinger["mid"]
+        informative[coin + "bb_upperband"] = bollinger["upper"]
+        informative[coin + "bb_width"] = (
+            informative[coin + "bb_upperband"] - informative[coin + "bb_lowerband"]
+        ) / informative[coin + "bb_middleband"]
+        informative[coin + "close-bb_lower"] = (
+            informative["close"] / informative[coin + "bb_lowerband"]
+        )
+
+        informative[coin + "roc"] = ta.ROC(informative, timeperiod=3)
+        informative[coin + "adx"] = ta.ADX(informative, window=14)
+
+        macd = ta.MACD(informative)
+        informative[coin + "macd"] = macd["macd"]
+        informative[coin + "pct-change"] = informative["close"].pct_change()
+        informative[coin + "relative_volume"] = (
+            informative["volume"] / informative["volume"].rolling(10).mean()
+        )
+
+        informative[coin + "pct-change"] = informative["close"].pct_change()
+
+        indicators = [col for col in informative if col.startswith(coin)]
+
+        for n in range(self.freqai_info["feature_parameters"]["shift"] + 1):
+            if n == 0:
+                continue
+            informative_shift = informative[indicators].shift(n)
+            informative_shift = informative_shift.add_suffix("_shift-" + str(n))
+            informative = pd.concat((informative, informative_shift), axis=1)
+
+        df = merge_informative_pair(df, informative, self.config["timeframe"], tf, ffill=True)
+        skip_columns = [(s + "_" + tf) for s in ["date", "open", "high", "low", "close", "volume"]]
+        df = df.drop(columns=skip_columns)
+
+        return df
+
+    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
+
+        # the configuration file parameters are stored here
+        self.freqai_info = self.config["freqai"]
+        self.pair = metadata['pair']
+
+        # the model is instantiated here
+        self.model = CustomModel(self.config)
```

Comment:
The issue with putting this in `bot_start()` is that it would carry around too much unnecessary data in ram. Imagine we are  training on 2 years of data for each of 30 pairs, your suggestion would keep all that data duplicated in ram. And what is worse, we may not even need that data again for a while if the user is simply inferencing a model saved to the drive. This class is just to help the strategy locate existing models to load them and make predictions with them, or to know whether or not it should train a new model and save it. As soon as we finish training/making a prediction, we can let python clean this up and move to the next pair because we have saved any relevant model/metadata to disk to be loaded later.

Correct me if my understanding of how the data would be held in ram is incorrect. But if it is true, then we still could move this line to `bot_loop_start()` in an effort to make the strategy "cleaner" for the user.

Reply:
i think you shouldn't think about these 2 (model loading, data loading) as attached / combined.

If i'm loading a model - i'd probably want to keep the model in memory - avoiding the "disk read" for the model.
now training is obviously different - you'll only want to load the data "when needed" - but i don't really see why the two can't be (at least partially) decoupled.

My point here is for live mode, mostly - if i assume i run a model on 20 pairs - we'd be loading 20 models for every pair on every iteration (or at least for every candle). 
Assuming that a trained model will not be humongous (it could be - but default models won't) - i would assume keeping that in memory would be fine?